File: Chess.Core/Position.cs
using System.Text.RegularExpressions;

namespace Chess.Core;

/// <summary>
/// <para>Structure to define a position.</para>
/// <para>Negative parameters represent unknowns or non-existing positions.</para>
/// </summary>
/// <param name="Row">Integer to represent the y-axis.</param>
/// <param name="Column">Integer to represent the x-axis.</param>
public readonly partial record struct Position(int Row, int Column)
{
    /// <summary>
    /// Turn positions into their chess notation equivalent.
    /// <list type="bullet">
    /// <item>
    /// In chess column/file is written before row/rank.
    /// </item>
    /// <item>
    /// rows start at 1 not 0.
    /// </item>
    /// <item>
    /// columns are written as chars a-z with a being the lowest (97 is the ascii offset for 'a' with there being 26 letters in total).
    /// </item>
    /// </list>
    /// </summary>
    /// <example>row: 0, column: 3 => "c1".</example>
    /// <returns>String representation of the current object.</returns>
    public override string ToString()
    {
        return $"{(char)((Column % 26) + 97)}{Row + 1}";
    }

    /// <summary>
    /// Get an instance of position from a position written in chess notation.
    /// </summary>
    /// <param name="dirtyPositionString">String in chess notation.</param>
    /// <returns>A position parsed from the string. Returns <c>new Position(-1, -1)</c> (or <c>"`0"</c> as string) if not parseable.</returns>
    public static Position FromString(string dirtyPositionString)
    {
        var cleanPositionString = dirtyPositionString.Trim().ToLower();
        var reg = PositionRegex();
        var isMatch = reg.IsMatch(cleanPositionString);

        return isMatch
            ? new Position(int.Parse(cleanPositionString[1..]) - 1, cleanPositionString[0] - 97)
            : new Position(-1, -1);
    }

    [GeneratedRegex(@"^[A-Z]\d+$", RegexOptions.IgnoreCase, "en-DE")]
    private static partial Regex PositionRegex();
}

File: Chess.Frontend/src/position.ts
export class Position{
  public row: number;
  public column: number;

  constructor(row: number, column: number) {
    this.row = row;
    this.column = column;
  }

  public equals(position: Position | null | undefined): boolean {
    if (position === null || position === undefined) return false;
    return (this.column === position.column && this.row === position.row);
  }

  public toString(): string {
    // Turn positions into their algebraic chess notation equivalent
    // Note: In chess column/file is written before row/rank.
    //       rows start at 1 not 0.
    //       columns are written as chars a-z with a being the lowest (97 is the ascii offset for 'a' with there being 26 letters in total)
    // E.g.: row: 0, column: 3 => "c1"
    return `${String.fromCharCode(this.column + 97)}${this.row + 1}`;
  }

  static fromString(input: string): Position{
    const match = input.match(/[a-z]\d+/);
    if (match != null && match.length >= 1) {
      const column = match[0].toUpperCase().charCodeAt(0) - 65;   // turn letters into numbers (A - Z => 0 - 26)
      const row = parseInt(match[0].slice(1)) - 1;                // turn one indexed number int zero indexed number (0 - 7 => 1 - 8)

      return new Position(row, column);
    }
    throw new Error("Unknown position");
  }
}
